### PR TITLE
Class names as template arguments

### PR DESCRIPTION
**Pull request by drwells**
This (mostly, perhaps completely) addresses #638, and also includes some small dependent followup patches.

I am not sure if we have a consensus on how to name these things. I picked `VectorType`, `QuadratureType`, etc., but these are easy to change to `VECTOR` or `QUADRATURE`. Should we vote on this? I have a slight preference towards `VectorType` but I think `VECTOR` is okay too.

### CODE REVIEW

**Comment by bangerth** [comment]
I like the `XyzType` convention. This looks good to me.

Of course this raises the question of whether we should replace `SP -> SparsityPatternType` and `DH -> DoFHandlerType` as well?

**Comment by drwells** [comment]
Yes, if we want to be consistent then we need to change the following as well:
* `SP` to `SparsityPatternType`
* `DH` to `DoFHandlerType`
* `VECTOR` to `VectorType`
* `STREAM` to `StreamType`
* `MATRIX` to `MatrixType`
* `POLY` to `PolynomialType`

and update the documentation (like the page on `MATRIX`) accordingly.

**Comment by drwells** [comment]
I am leaving this list here mostly as a reminder for myself as to what needs to be done.

- [x] `BlockMatrix` to `BlockMatrixType`
- [ ] `BLOCK_VECTOR` to `BlockVectorType`
- [x] `FEValues` to `FEValuesType`
- [x] `MATRIX` to `MatrixType`
- [ ] `PRECONDITION` to `PreconditionerType`
- [x] `Quadrature` to `QuadratureType`
- [ ] `SOLVER` to `SolverType`
- [ ] `SP` to `SparsityPatternType`
- [x] `Vector` to `VectorType`
- [x] `VECTOR` to `VectorType`
- [x] `VECTOR` to `VectorType`

**Comment by drwells** [comment]
This PR is getting rather huge so I removed the `VECTOR` to `VectorType` et al renames (that will be a future PR). This should be good to go (and it should close #638).

**Comment by bangerth** [comment]
Last I looked at this, this looked good to me. I'm traveling this week but maybe someone else could take a look?

File: include/deal.II/grid/grid_refinement.h [hunk 1]
```diff
@@ -146,14 +146,14 @@ namespace GridRefinement
...
+  template <int dim, class VectorType, int spacedim>
   void
-  refine_and_coarsen_fixed_number (
-    Triangulation<dim,spacedim> &tria,
-    const Vector                &criteria,
-    const double                top_fraction_of_cells,
-    const double                bottom_fraction_of_cells,
-    const unsigned int          max_n_cells = std::numeric_limits<unsigned int>::max());
+  refine_and_coarsen_fixed_number
+  (Triangulation<dim,spacedim> &tria,
```
Comment (Rombur) [inline include/deal.II/grid/grid_refinement.h]: Any reason why you changed the line the bracket is on? Is it consistent with what is done elsewhere?

File: include/deal.II/grid/grid_refinement.h [hunk 2]
```diff
@@ -207,14 +207,14 @@ namespace GridRefinement
...
+  template <int dim, class VectorType, int spacedim>
   void
-  refine_and_coarsen_fixed_fraction (
-    Triangulation<dim,spacedim> &tria,
-    const Vector                &criteria,
-    const double                top_fraction,
-    const double                bottom_fraction,
-    const unsigned int          max_n_cells = std::numeric_limits<unsigned int>::max());
+  refine_and_coarsen_fixed_fraction
+  (Triangulation<dim,spacedim> &tria,
```
Comment (Rombur) [inline include/deal.II/grid/grid_refinement.h]: Same here

File: source/dofs/dof_tools_sparsity.cc [hunk 3]
```diff
@@ -328,12 +327,12 @@ namespace DoFTools
...
 
-  template <class DH, class SparsityPattern>
+  template <class DH, class SP>
   void
-  make_boundary_sparsity_pattern (
-    const DH                        &dof,
-    const std::vector<types::global_dof_index> &dof_to_boundary_mapping,
-    SparsityPattern                 &sparsity)
+  make_boundary_sparsity_pattern
+  (const DH                                   &dof,
```
Comment (Rombur) [inline source/dofs/dof_tools_sparsity.cc]: Same as before.

File: source/dofs/dof_tools_sparsity.cc [hunk 4]
```diff
@@ -397,12 +396,12 @@ namespace DoFTools
...
 
-  template <class DH, class SparsityPattern>
-  void make_boundary_sparsity_pattern (
-    const DH                                        &dof,
-    const typename FunctionMap<DH::space_dimension>::type &boundary_ids,
-    const std::vector<types::global_dof_index>                 &dof_to_boundary_mapping,
-    SparsityPattern                                 &sparsity)
+  template <class DH, class SP>
+  void make_boundary_sparsity_pattern
+  (const DH                                              &dof,
```
Comment (Rombur) [inline source/dofs/dof_tools_sparsity.cc]: Same as before.

File: source/hp/fe_values.cc [hunk 5]
```diff
@@ -25,12 +25,12 @@ namespace internal
...
 
-    template <int dim, int q_dim, class FEValues>
-    FEValuesBase<dim,q_dim,FEValues>::
-    FEValuesBase (const dealii::hp::MappingCollection<dim,FEValues::space_dimension> &mapping_collection,
-                  const dealii::hp::FECollection<dim,FEValues::space_dimension>      &fe_collection,
-                  const dealii::hp::QCollection<q_dim>     &q_collection,
-                  const UpdateFlags                   update_flags)
+    template <int dim, int q_dim, class FEValuesType>
+    FEValuesBase<dim,q_dim,FEValuesType>::FEValuesBase
+    (const dealii::hp::MappingCollection<dim,FEValuesType::space_dimension> &mapping_collection,
```
Comment (Rombur) [inline source/hp/fe_values.cc]: Same as before.

File: source/hp/fe_values.cc [hunk 6]
```diff
@@ -45,14 +45,14 @@ namespace internal
...
 
 
-    template <int dim, int q_dim, class FEValues>
-    FEValuesBase<dim,q_dim,FEValues>::
-    FEValuesBase (const dealii::hp::FECollection<dim,FEValues::space_dimension>      &fe_collection,
-                  const dealii::hp::QCollection<q_dim> &q_collection,
-                  const UpdateFlags         update_flags)
+    template <int dim, int q_dim, class FEValuesType>
+    FEValuesBase<dim,q_dim,FEValuesType>::FEValuesBase
+    (const dealii::hp::FECollection<dim,FEValuesType::space_dimension> &fe_collection,
```
Comment (Rombur) [inline source/hp/fe_values.cc]: Same as before.

File: source/hp/fe_values.cc [hunk 7]
```diff
@@ -66,12 +66,12 @@ namespace internal
...
-    template <int dim, int q_dim, class FEValues>
-    FEValues &
-    FEValuesBase<dim,q_dim,FEValues>::
-    select_fe_values (const unsigned int fe_index,
-                      const unsigned int mapping_index,
-                      const unsigned int q_index)
+    template <int dim, int q_dim, class FEValuesType>
+    FEValuesType &
+    FEValuesBase<dim,q_dim,FEValuesType>::select_fe_values
+    (const unsigned int fe_index,
```
Comment (Rombur) [inline source/hp/fe_values.cc]: Here too.

**Comment by drwells** [comment]
@Rombur good question. I thought that ending a line with an opening parenthesis was less common, but now that I actually checked I see that (on `master`)

```
[drwells@arch include]$ ack '^ *\(.*,$' | wc -l
294
```

and 

```
[drwells@arch include]$ ack '\($' | wc -l    
303
```

which implies otherwise. Does anyone have a strong opinion about which of these styles is preferable?

**Comment by Rombur** [comment]
Since both are used almost as often as each either, I'll merge this.

**Comment by bangerth** [comment]
Different people had different styles over the years in dealing with long function names and following arguments. I personally put the opening parenthesis on the next line, I think Guido usually put it after the function name and the arguments on the next line. I don't think it matters one way or the other (though I nonetheless occasionally convert some when I'm there anyway).

**Pull request merged**